### PR TITLE
Tic-tac-toe command

### LINKED ISSUE #300: Tic-Tac-Toe command
`.ttt` or `.tictactoe` command lets users play tic-tac-toe using reactions, either player vs player or player vs AI-like program.

## Implementation details

List the changes required for the implementation here.

- [ ] Write the bot command (a new cog)
- [x] Draw beautiful x, o emojis
- [x] Add required emojis
- [ ] Add the command to our README.md

## Additional information

Emojis have been created:

```
:empty_square: - 632278399296995361
:x_square - 632278427260682281
:o_square - 632278452413661214
```

x
o

### PR DESCRIPTION
**Pull request by ks129**
## Relevant Issues

Closes #300 

## Description

I added `tictactoe` command like described in issue, + added `tictactoe history|log` to show a list of current session games and `tictactoe history|log show <id>` to show the board of the game.

## Reasoning

Because there was approved issue.

## Screenshots

Screenshot 2020-04-12 at 19 22 37
Screenshot 2020-04-12 at 19 22 46
Screenshot 2020-04-12 at 19 22 53
Screenshot 2020-04-12 at 19 23 01

## Did you:

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] If dependencies have been added or updated, run `pipenv lock`?
- [x] **Lint your code** (`pipenv run lint`)?
- [x] Set the PR to **allow edits from contributors**?

### CODE REVIEW

**Review by kosayoda — CHANGES_REQUESTED** [review]
Thank you for the pull request, and sorry for the delay in review. I've done a quick first round of review, which involves some high level changes. Feel free to ask for clarification.

PS: Can you also add a spacer to the first line of the board whenever it is printed? Using the `compact` display setting, the game looks like this:
image

**Comment by Akarys42** [comment]
Hey @ks129, are you still on this?

**Comment by ks129** [comment]
@Akarys42 Yes, just after finished these Dewikification PRs I didn't have time to apply these changes because of school.

**Comment by ks129** [comment]
@kosayoda Review parts is now applied, but I didn't added this newline for compact mode, because this will make emojis sooo small in *normal* mode what makes this hardly readable.

**Comment by Xithrius** [comment]
@ks129 Hello, what's the status of this PR? 

If you happen to need an extra set of eyes/hands for reviewing/writing code, I'm down.

**Comment by ks129** [comment]
@Xithrius This PR is waiting review.

**Comment by Xithrius** [comment]
> @Xithrius This PR is waiting review.

I will review once I wake. Thank you for the response, have a good one!

**Comment by Xithrius** [comment]
Sorry, completely forgot about this PR. I will test things after reviewing the code itself.

File: bot/exts/evergreen/tic_tac_toe.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+            or board[4] == board[5] == board[6]
+            or board[7] == board[8] == board[9]
+            # Vertical
+            or board[1] == board[4] == board[7]
+            or board[2] == board[5] == board[8]
+            or board[3] == board[6] == board[9]
+            # Diagonal
+            or board[1] == board[5] == board[9]
+            or board[3] == board[5] == board[7]
+    ):
```
Comment (Xithrius) [inline bot/exts/evergreen/tic_tac_toe.py:32]: What do you think about the suggestion below to make the code cleaner to read?
```suggestion
    if any((
            # Horizontal
            board[1] == board[2] == board[3],
            board[4] == board[5] == board[6],
            board[7] == board[8] == board[9],
            # Vertical
            board[1] == board[4] == board[7],
            board[2] == board[5] == board[8],
            board[3] == board[6] == board[9],
            # Diagonal
            board[1] == board[5] == board[9],
            board[3] == board[5] == board[7]
    )):
```

**Review by Xithrius — COMMENTED** [review]
Looks good to me besides one possible change.

I will begin testing now.

EDIT: Everything looks good, but I think a solution should be found where this game can run in compact mode. Maybe we can use an embed?

File: bot/exts/evergreen/tic_tac_toe.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    def __init__(self, bot: SeasonalBot):
+        self.bot = bot
+        self.games: t.List[Game] = []
+
+    @guild_only()
+    @is_channel_free()
+    @is_requester_free()
+    @group(name="tictactoe", aliases=("ttt",), invoke_without_command=True)
+    async def tic_tac_toe(self, ctx: Context, opponent: t.Optional[discord.User]) -> None:
+        """Tic Tac Toe game. Play agains friends or AI. Use reactions to add your mark to field."""
```
Comment (Xithrius) [inline bot/exts/evergreen/tic_tac_toe.py:251]:
```suggestion
       """Tic Tac Toe game. Play against friends or AI. Use reactions to add your mark to field."""
       ```

**Comment by ks129** [comment]
@Xithrius Board is now moved to embed.

File: bot/exts/evergreen/tic_tac_toe.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+import discord
+from discord.ext.commands import Cog, Context, check, group, guild_only
+
+from bot.bot import Bot
+from bot.constants import Emojis
+from bot.utils.pagination import LinePaginator
+
+CONFIRMATION_MESSAGE = (
+    "{opponent}, {requester} want to play Tic-Tac-Toe against you. React to this message with "
```
Comment (ChrisLovering) [inline bot/exts/evergreen/tic_tac_toe.py:13]:
```suggestion
    "{opponent}, {requester} wants to play Tic-Tac-Toe against you. React to this message with "
```

File: bot/exts/evergreen/tic_tac_toe.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    def format_board(self) -> str:
+        """Get formatted tic-tac-toe board for message."""
+        board = list(self.board.values())
+        return "\n".join(
+            (f"{board[line]} {board[line + 1]} {board[line + 2]}" for line in range(0, len(board), 3))
+        )
+
+    async def play(self) -> None:
+        """Start and handle game."""
+        await self.ctx.send("It's time for game! Let's begin.")
```
Comment (ChrisLovering) [inline bot/exts/evergreen/tic_tac_toe.py:190]: Maybe `time for a game` / `time for the game` reads better here?

Comment (ks129) [reply bot/exts/evergreen/tic_tac_toe.py:190]: Not sure about it. I don't know these articles stuff well.

File: bot/exts/evergreen/tic_tac_toe.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+        board = await self.ctx.send(
+            embed=discord.Embed(description=self.format_board())
+        )
+        await self.add_reactions(board)
+
+        for _ in range(9):
+            if isinstance(self.current, Player):
+                announce = await self.ctx.send(f"{self.current.user.mention}, your turn! React to emoji to mark field.")
```
Comment (ChrisLovering) [inline bot/exts/evergreen/tic_tac_toe.py:198]: Maybe `, it's your turn! React with an emoji to take your go` reads better?

File: bot/exts/evergreen/tic_tac_toe.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+            self.board[pos] = self.current.symbol
+            await board.edit(
+                embed=discord.Embed(description=self.format_board())
+            )
+            await board.clear_reaction(Emojis.number_emojis[pos])
+            if check_win(self.board):
+                self.winner = self.current
+                self.loser = self.next
+                await self.ctx.send(
+                    f":tada: {self.current} is won this game! :tada:"
```
Comment (ChrisLovering) [inline bot/exts/evergreen/tic_tac_toe.py:216]:
```suggestion
                    f":tada: {self.current} has won this game! :tada:"
```

File: bot/exts/evergreen/tic_tac_toe.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+                )
+                await board.clear_reactions()
+                break
+            self.current, self.next = self.next, self.current
+        if not self.winner:
+            self.draw = True
+            await self.ctx.send("It's DRAW!")
```
Comment (ChrisLovering) [inline bot/exts/evergreen/tic_tac_toe.py:223]:
```suggestion
            await self.ctx.send("It's a DRAW!")
```

File: bot/exts/evergreen/tic_tac_toe.py (hunk 6 above)
Comment (ks129) [reply bot/exts/evergreen/tic_tac_toe.py:216]: Maybe `{user} won this game!` even better? Has doesn't sound well in this context.

Comment (ChrisLovering) [reply bot/exts/evergreen/tic_tac_toe.py:216]: Yea, that sounds good too.

**Comment by Xithrius** [comment]
> @Xithrius Board is now moved to embed.

Alright. I'll test it myself as soon as you mark this as ready for a review.

**Comment by ks129** [comment]
@Xithrius What do you mean? This isn't Draft PR.

**Comment by Xithrius** [comment]
Sorry, miscommunication I my end, I meant whenever you wanted me to review (marking it ready for me).

File: bot/exts/evergreen/tic_tac_toe.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+    @tic_tac_toe_logs.command(name="show", aliases=("s",))
+    async def show_tic_tac_toe_board(self, ctx: Context, game_id: int) -> None:
+        """View game board by ID (ID is possible to get by `.tictactoe history`)."""
+        if len(self.games) < game_id:
+            await ctx.send("Game don't exist.")
+            return
+        game = self.games[game_id - 1]
+        await ctx.send(
+            f"{game.winner} :trophy: vs {game.loser}"
+        )
```
Comment (Xithrius) [inline bot/exts/evergreen/tic_tac_toe.py:319]: I don't think this line should be multiple when every other line in this file has `await ctx.send()` as one line.
```suggestion
       await ctx.send(f"{game.winner} :trophy: vs {game.loser}")
```

**Review by Xithrius — CHANGES_REQUESTED** [review]

**Review by Xithrius — APPROVED** [review]
I tested this on my own, looks good to me.

File: bot/exts/evergreen/tic_tac_toe.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+            """Check does user who reacted is user who we want, message is board and emoji is in board values."""
+            return (
+                u.id == self.user.id
+                and msg.id == r.message.id
+                and r.emoji in board.values()
+                and r.emoji in Emojis.number_emojis.values()
+            )
+
+        try:
+            react, _ = await self.ctx.bot.wait_for('reaction_add', timeout=120.0, check=check_for_move)
```
Comment (RohanJnr) [inline bot/exts/evergreen/tic_tac_toe.py:61]: I think having this set to something lower like 30s for the timeout should be better as this is not a game where we would spend a lot of time thinking

**Review by RohanJnr — APPROVED** [review]

**Pull request merged**